pkg/integrations/ai_agent/ai_agent_integration.go

```diff
@@ -516,3 +522,153 @@ func (e *AIAgentExecutorV2) GetNodeIDFromOutputID(outputID string) string {
+	}
+
+	// Record execution history
+	if executionContext.HistoryRecorder != nil {
+		e.recordExecutionHistory(ctx, executeParams, agentSettings, executedToolNodeIDs, executionContext.HistoryRecorder)
+	}
+
+	return nil
+}
+
+
+func (e *AIAgentExecutorV2) publishNodeEvents(ctx context.Context, nodeID, workflowID, executionID string) {
+	// Publish node execution started event
+	startedEvent := &domain.NodeExecutionStartedEvent{
+		WorkflowID:          workflowID,
+		WorkflowExecutionID: executionID,
+		NodeID:              nodeID,
+		Timestamp:           time.Now().UnixNano(),
+	}
+
+	err := e.eventPublisher.PublishEvent(ctx, startedEvent)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failed to publish node execution started event for node %s", nodeID)
+		return
+	}
+
+	// Publish node execution completed event
+	completedEvent := &domain.NodeExecutedEvent{
+		WorkflowID:          workflowID,
+		WorkflowExecutionID: executionID,
+		NodeID:              nodeID,
+		Timestamp:           time.Now().UnixNano(),
+		ItemsByInputID:      map[string]domain.NodeItems{},
+		ItemsByOutputID:     map[string]domain.NodeItems{},
+		ExecutionOrder:      0, // Connected nodes don't have execution order
+	}
+
+	err = e.eventPublisher.PublishEvent(ctx, completedEvent)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failed to publish node execution completed event for node %s", nodeID)
+	}
+}
+
+func (e *AIAgentExecutorV2) recordExecutionHistory(ctx context.Context, executeParams ExecuteParams, agentSettings AgentSettings, executedToolNodeIDs map[string]bool, recorder domain.ExecutionHistoryRecorder) {
+	if executeParams.LLM != nil {
+		e.recordNodeExecutionHistory(executeParams.LLM.NodeID, "openai", "chat_completion", recorder)
+	}
+
+	if executeParams.Memory != nil {
+		e.recordNodeExecutionHistory(executeParams.Memory.NodeID, "flowbaker_agent_memory", "store_conversation", recorder)
+	}
+
+	// Try to get tool execution details from tracker first
+	execContext, hasContext := domain.GetWorkflowExecutionContext(ctx)
+	if hasContext && execContext.ToolTracker != nil {
+		execContext.ToolTracker.ForEachExecution(func(nodeID string, toolExec *domain.ToolExecution) {
+			e.recordNodeExecutionHistory(nodeID, string(toolExec.Identifier.IntegrationType), string(toolExec.Identifier.ActionType), recorder)
+		})
+	} else {
+		// Fallback to legacy method
+		for nodeID := range executedToolNodeIDs {
+			for _, tool := range agentSettings.Tools {
+				if tool.NodeID == nodeID {
+					e.recordNodeExecutionHistory(nodeID, string(tool.IntegrationType), string(tool.WorkflowNode.ActionType), recorder)
+					break
+				}
+			}
+		}
+	}
+}
+
+func (e *AIAgentExecutorV2) recordNodeExecutionHistory(nodeID string, integrationType string, actionType string, recorder domain.ExecutionHistoryRecorder) {
+	now := time.Now()
+	
+	recorder.AddNodeExecution(domain.NodeExecution{
+		ID:                     fmt.Sprintf("%s_%d", nodeID, now.UnixNano()),
```

Comment:
we should check workflow node execution ids i think this is not the true way for generating ids